### ISSUE TITLE
Understanding File Size and Inference with onnx export

### ISSUE THREAD

**Issue by rhino-098** [Xilinx/brevitas#1346]
Hi everyone,

I'm working with the ONNX export feature and have a few questions about the expected behavior to make sure I'm on the right track.

My main observation is that when I change the weight_bit_width (8bits, 4bits, 2bits)  in a QuantLinear layer, the final .onnx file size remains constant. Is this the intended behavior?

This leads to my next question about inference. When a backend like ONNX Runtime executes the model on a standard CPU or GPU, is the matrix multiplication performed using low-bit integer arithmetic? Or do the quantized values get converted back to float32 first?

I'm also curious about where to expect performance gains. Are the speed improvements from quantization primarily seen on specialized hardware like FPGAs with a toolchain like FINN, or can we also expect significant speedups on modern CPUs and GPUs when using Brevitas?

I'm trying to understand where the primary benefits of quantization, such as model size reduction and faster inference, are meant to appear in the workflow. Am I missing a common step after using export_onnx_qcdq, or are these benefits only realized during execution, for instance, when running the exported ONNX model with ONNX Runtime?

Thanks in advance for any clarification.

For context, here's the simple script I'm using:

```python

from brevitas.quant import Int8ActPerTensorFloat
from brevitas.export import export_onnx_qcdq
import torch
from brevitas.nn import QuantLinear

float_input = torch.randn(3, 5000)

quant_linear = QuantLinear(
    5000, 10000,
    weight_bit_width=8, # Changing this doesn't affect file size
    bias=False)

export_model = export_onnx_qcdq(
    quant_linear,
    args=float_input,
    export_path="example.onnx")

**Comment by Giuseppe5** [comment]
> My main observation is that when I change the weight_bit_width (8bits, 4bits, 2bits) in a QuantLinear layer, the final .onnx file size remains constant. Is this the intended behavior?

With respect to compression, Brevitas' ONNX export by default will compress the weights, although the only possibility is to map to int8. If you quantize to int4, the weights will still be stored in `int8` containers, so you won't see an effective size reduction.
You can check that yourself in your ONNX graph. The weights should be in their integer representation, with only a DequantizeLinear (and eventually a clip for sub int8 quantization) operation.

> This leads to my next question about inference. When a backend like ONNX Runtime executes the model on a standard CPU or GPU, is the matrix multiplication performed using low-bit integer arithmetic? Or do the quantized values get converted back to float32 first?

During Pytorch execution, and also when exporting to ONNX QCDQ, we adopt what is usually referred as "Fake Quantization". As you point out, everything gets reconverted to float32 before execution. Having a ONNX QCDQ graph, you would have to re-arrange quant operations and have dedicated and optimized quantized kernel to take advantage of faster integer computation. 

> I'm also curious about where to expect performance gains. Are the speed improvements from quantization primarily seen on specialized hardware like FPGAs with a toolchain like FINN, or can we also expect significant speedups on modern CPUs and GPUs when using Brevitas?

In general, Brevitas is not concerned with the "performance" part of the quantization, but rather trying with the simulation and accuracy preservation. The reasons for this are multiple and somewhat difficult to summarize but to greatly oversimply, there are too many different combinations of hardware and quantization configurations to expect that Brevitas could offer competitive performance for all of them. 
We provide export flows to different compiler infrastructure, like [Shark-AI](https://xilinx.github.io/brevitas/dev/user_guide/export_shark.html), [GGUF](https://xilinx.github.io/brevitas/dev/user_guide/export_gguf.html), and FINN and you mentioned earlier.

We are interested in expanding our support for other compiler flows, but in terms of priority it is not as high as research in better accuracy preservation methods.



Hope this answer all your questions but please feel free to reach out if something is still not clear

**Comment by rhino-098** [comment]
Thanks a lot for such a detailed and insightful response, it clarifies a lot.

Other than the simulation and accuracy aspects, does Brevitas offer any improvements in terms of model size when exporting to ONNX and using that exported model for inference? For instance, even though the computations happen in floating point, does quantization still help with enabling the loading or deployment of larger models that might otherwise be too memory-heavy in full-precision?

Also, is there any difference in inference behavior, performance, or observed accuracy between running a Brevitas quantized model natively in PyTorch (without exporting) and running the exported ONNX model using ONNX Runtime?

Thanks again for your time and support.
